File: backend/utils/emi_calculator.py
from decimal import Decimal

def calculate_emi_and_payable(principal, annual_rate, duration_months):
    p = Decimal(str(principal))
    r = Decimal(str(annual_rate))
    n = duration_months

    if r == 0:
        emi = p / n
        total_payable = p
        return round(emi, 2), round(total_payable, 2)

    mr = (r / Decimal('12')) / Decimal('100')
    one_plus_r_n = (Decimal('1') + mr) ** n
    emi = p * mr * one_plus_r_n / (one_plus_r_n - Decimal('1'))
    total_payable = emi * n
    return round(emi, 2), round(total_payable, 2)

File: frontend/src/utils/emiCalculator.js
export function calculateEMIAndPayable(principal, annualRate, durationMonths) {
  const p = parseFloat(principal) || 0
  const r = parseFloat(annualRate) || 0
  const n = parseInt(durationMonths) || 0

  if (p <= 0 || n <= 0) return { emi: 0, totalPayable: 0, totalInterest: 0 }

  if (r === 0) {
    const emi = p / n
    return { emi: round2(emi), totalPayable: round2(p), totalInterest: 0 }
  }

  const mr = (r / 12) / 100
  const onePlusR_n = Math.pow(1 + mr, n)
  const emi = p * mr * onePlusR_n / (onePlusR_n - 1)
  const totalPayable = emi * n
  const totalInterest = totalPayable - p

  return {
    emi: round2(emi),
    totalPayable: round2(totalPayable),
    totalInterest: round2(totalInterest)
  }
}

function round2(val) {
  return Math.round(val * 100) / 100
}
